uarray/_uarray.pyi

```diff
@@ -0,0 +1,126 @@
+"""Annotations for the ``uarray._uarray`` extension module."""
+
+import types
+from collections.abc import Callable, Iterable
+from typing import Any, final, overload
+
+import uarray
+from uarray._typing import (
+    _PyGlobalDict,
+    _PyLocalDict,
+    _ReplacerFunc,
+    _SupportsUAConvert,
+    _SupportsUADomain,
+)
+
+class BackendNotImplementedError(NotImplementedError): ...
+
+@final
+class _SkipBackendContext:
+    def __init__(self, backend: _SupportsUADomain) -> None: ...
+    def __enter__(self) -> None: ...
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: types.TracebackType | None,
+        /,
+    ) -> None: ...
+    def _pickle(self) -> tuple[_SupportsUADomain]: ...
+
+@final
+class _SetBackendContext:
+    def __init__(
+        self,
+        backend: _SupportsUADomain,
+        coerce: bool = ...,
+        only: bool = ...,
+    ) -> None: ...
+    def __enter__(self) -> None: ...
+    def __exit__(
+        self,
+        exc_type: type[BaseException] | None,
+        exc_value: BaseException | None,
+        traceback: types.TracebackType | None,
+        /,
+    ) -> None: ...
+    def _pickle(self) -> tuple[_SupportsUADomain, bool, bool]: ...
+
+# NOTE: Parametrize w.r.t. `__ua_domain__` when returning, but use `Any`
+# when used as argument type. Due to lists being invariant the `__ua_domain__`
+# protocol will likelly be disruptivelly strict in the latter case, hence the
+# usage of `Any` as an escape hatch.
+@final
+class _BackendState:
+    def _pickle(self) -> tuple[
+        _PyGlobalDict[_SupportsUADomain],
+        _PyLocalDict[_SupportsUADomain],
+        bool,
+    ]: ...
+    @classmethod
+    def _unpickle(
+        cls,
+        py_global: _PyGlobalDict[Any],
+        py_locals: _PyLocalDict[Any],
+        use_thread_local_globals: bool,
+        /,
+    ) -> _BackendState: ...
+
+@final
+class _Function:
+    def __init__(
+        self,
+        extractor: Callable[..., tuple[uarray.Dispatchable, ...]],
+        replacer: None | _ReplacerFunc,
+        domain: str,
+        def_args: tuple[Any, ...],
+        def_kwargs: dict[str, Any],
+        def_impl: None | Callable[..., Any],
+    ) -> None: ...
+    def __repr__(self) -> str: ...
+    def __call__(self, *args: Any, **kwargs: Any) -> Any: ...
```

Comment:
Would there be a way to make this have the "same input signature" as the `extractor`? I'm assuming not, but it's worth asking.

Reply:
It's a good thing you asked in that case, because the answer is "yes". Initially I had some trouble figuring out the signature of `__call__`, which is why I left it as broad as it is now.

Reply:
Please feel free to ask if there are other things that are left unspecified. Very willing to help move this along.